rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,318 @@
+they target `cloud-stable` when the cluster is configured for continuous agent upgrades.
+
+The current implementation of the `node-join` script template relies upon direct downloads from `get.gravitational.com`. It would
+be preferable to force use of a package manager when installing `cloud-stable`, as that would let us improve standardization, minimize
+the amount of custom logic in the download path, and (most importantly) reduce the complexity of the periodic upgrade operation,
+which is the most sensitive part of this proposal.
+
+I believe the best possible outcome would be to unify the core logic of the "discovery", "join", and "upgrade" scripts into a single
+cross-platform function so that we have a single source of truth for package manager discovery and invocation going forward. The
+feasability of this will require further investigation.
+
+
+### Summary of Changes
+
+- Agents will expose a flags to enable the VM and k8s variants of the agent upgrade behavior.
+
+- Package repositories will expose a `cloud-stable` variant of packages which will lag slightly behind "latest"
+and be managed indipendently of the standard release process.
+
+- A new `cloud-stable` helm chart (or option) will be dstributed.
+
+- A new `teleport-upgrade` script and accompanying systemd timer unit will be distributed as part of the
+`cloud-stable` apt/yum packages.
+
+- Teleport auth servers will be modified to be able to serve the maintenance window to agents.
+
+- Cloud clusters will have the option to track `cloud-stable` rather than a specific major version.
+
+- Teleport discovery and join features will be modified to serve appropriate install scripts when `cloud-stable`
+tracking is enabled for a cluster.
+
+
+### Security Considerations
+
+For the most part, this proposal relies on the existing security model of our package distribution system. So long
+as package managers/repos remains secure, this system aught to be secure by extension, and as we improve the security
+of our distribution systems (e.g. adding image signing to OCI repos), we should ensure that this system benefits from
+those improvements. The only truly novel components
+are the upgrade script and timer unit. Unauthorized modification would be bad, but that is just a matter of setting
+correct file permissions. I think the main risk induced by these additions would be in making the upgrade script too
+dynamic. Upgrade scripts should not attempt multiple different download options (e.g. an upgrade script
+installed by `yum` shouldn't fallback to invoking `apt` if something in the host environment changes, or if an unexpected
+error occurs). Upgrade scripts should also not accept any variables/inputs from auth or from teleport. The origin of the
+upgrade script should statically determine its behavior.
+
+I'm still in the early stages of thinking over attack vectors, so please comment if anything comes time mind!
+
+
+### Future Considerations
+
```

Comment:
Depending on our implementation details we could potentially add canary deployments in the future to minimize the impact of bad updates. The `cloud-stable` channel could provide different versions based upon some customer identifier.